lib/src/rules/implicit_reopen.dart

```diff
@@ -0,0 +1,167 @@
+
+**BAD:**
+```dart
+interface class I {}
+
+class C extends I {}
+```
+
+**GOOD:**
+```dart
+interface class I {}
+
+final class C extends I {}
+```
+
+```dart
+import 'package:meta/meta.dart';
+
+interface class I {}
+
+@reopen
+class C extends I {}
+```
+''';
+
+class ImplicitReopen extends LintRule {
+  static const LintCode code = LintCode('implicit_reopen',
+      "The {0} '{1}' reopens '{2}' because it is not marked '{3}'",
+      correctionMessage:
+          "Try marking '{1}' '{3}' or annotating it with '@reopen'");
+
+  ImplicitReopen()
+      : super(
+            name: 'implicit_reopen',
+            description: _desc,
+            details: _details,
+            state: State.experimental(),
+            group: Group.errors);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addClassDeclaration(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  void checkElement(InterfaceElement? element, NamedCompilationUnitMember node,
+      {required String type}) {
+    if (element == null) return;
+    if (element.hasReopen) return;
+    if (element.isSealed) return;
+
+    var library = element.library;
+    var supertype = element.superElement;
+    if (supertype.library != library) return;
+
+    if (element.isBase) {
+      if (supertype.isFinal) {
+        reportLint(node,
+            target: element, other: supertype!, reason: 'final', type: type);
+        return;
+      } else if (supertype.isInterface) {
+        reportLint(node,
+            target: element,
+            other: supertype!,
+            reason: 'interface',
+            type: type);
+        return;
+      }
+    } else if (element.hasNoModifiers) {
+      if (supertype.isInterface) {
+        reportLint(node,
+            target: element,
+            other: supertype!,
+            reason: 'interface',
+            type: type);
+        return;
+      }
+    }
+  }
+
+  void reportLint(
+    NamedCompilationUnitMember member, {
+    required String type,
+    required InterfaceElement target,
+    required InterfaceElement other,
+    required String reason,
+  }) {
+    rule.reportLintForToken(member.name,
+        arguments: [type, target.name, other.name, reason]);
+  }
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    checkElement(node.declaredElement, node, type: 'class');
+  }
+}
+
+extension on InterfaceElement? {
+  bool get hasNoModifiers => !isInterface && !isBase && !isSealed && !isFinal;
+
+  bool get isBase {
+    var self = this;
+    if (self is ClassElement) return self.isBase;
+    if (self is MixinElement) return self.isBase;
+    return false;
+  }
+
+  bool get isFinal {
+    var self = this;
+    if (self is ClassElement) return self.isFinal;
+    if (self is MixinElement) return self.isFinal;
```

Comment:
I would avoid checking if `MixinElement` has any non-base modifier. I will be removing these soon and this would add extra clean up for you :)

Reply:
Cool!  That got me thinking about mixin classes too (on which were overreporting) so double-thanks!